ecopod/sendRawTransactionConditional.md

```diff
@@ -0,0 +1,53 @@
+
+**Note:** As a result of [EIP-3074](https://eips.ethereum.org/EIPS/eip-3074), the interop project is also exploring external validation as a mechanism to preventing DoS attack. A low-stakes scoped version of this for `eth_sendRawTransactionConditional` is a good opportunity for learning.
+
+
+# Problem Statement + Context
+
+Bundlers aggregate UserOps from an external mempool into a single transaction sent to the enshrined 4337 `Entrypoint` contract. It is the Bundler's responsibility to ensure every UserOp in the transaction will not revert when its validation step is executed. In between tx submission and inclusion, there is a chance network state changes causes the transaction to revert, even after local bundler validation.
+
+With private Bundler mempools, the likelihood of these reverts are small. However shared 4337 mempools are coming into [production](https://medium.com/etherspot/decentralized-future-erc-4337-shared-mempool-launches-on-ethereum-b6c860072f41), launched on Ethereum and some L2 testnets, decentralizing 4337 infrastructure. This is possible with no change on Ethereum through special block builders like Flashbots which support atomic transaction inclusion. An endpoint like `eth_sendRawTransactionConditional` is required to launch a shared mempools on L2 as there's only a single block builder, the sequencer, and the likelihood of reverts with a shared mempool would be too high of a cost for bundlers to operate.
+
+
+# Alternatives Considered
+
+## Continue with private 4337 mempools
+
+4337 account abstraction is live on optimism. Dapps likely utilize the bundler endpoints that are come with a vendor-specific SDK -- Alchemy, Pimilico, Thirdweb, etc. As 4337 infrastructure becomes more permissionless, we will likely later have to play catch-up to ensure the OP-Stack remains compatible while other L2 offerings have already moved towards supporting `eth_sendRawTransactionConditional`.
+
+## Verticalize the OP-Stack
+
+Rather than externalize the 4337 mempool, the op-stack could natively offer a UserOp mempool alongside the regular mempool, avoiding conflicts as the sequencer can ensure the bundled UserOps do not conflict with the latest chain state when sealing a new block. However, this adds additional complexity to the stack, where the proposer-builder seperation in 4337 nicely keeps account abstraction out of protocol. A chain operator can still choose to verticalize in proposed solution solution by allowlisting only its own bundler through `eth_sendRawTransactionConditional`, achieving the same outcome.
+
+
+# Proposed Solution
+
+1. Implement `eth_sendRawTransactionConditional` as described in the [spec](https://notes.ethereum.org/@yoav/SkaX2lS9j). A draft implementation for this [exists](https://github.com/ethereum/go-ethereum/compare/master...tynes:go-ethereum:eip4337)today. The conditional is applied on submission and tx inclusion.
```

Comment:
Can you define submission and tx inclusion? This isn't precise enough to really know. The invariants are checked on entrance to the mempool and also during block building

Reply:
For things that may cause denial of service, we want to be able to solve the problem with horizontal scalability. This leads me to believe a good longer term solution is a more generic "superchain backend" that can help with these invariants. I think `revm` + `EthersDB` (think foundry forking) is a really good way to horizontally scale here. Its also possible to build this on geth using a `StateDB` implementation that can fetch state from RPC

Reply:
yea that's a good point, I added a section in `Optimized State Lookups` as areas for improvement in the proposed solution